File: src/python/xerxes/session/summarizer.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Cheap heuristic summaries of recorded sessions.

:class:`SessionSummarizer` produces a :class:`SessionSummary` from a
:class:`SessionRecord` using string-only heuristics: title from the first
prompt, a one-paragraph synopsis, distinct tool names, success/failure
outcome. An optional LLM callback can refine the synopsis without inventing
new facts — failures fall back to the heuristic draft.
"""

from __future__ import annotations

import logging
import typing as tp
from dataclasses import asdict, dataclass, field

from .models import SessionRecord

logger = logging.getLogger(__name__)


@dataclass
class SessionSummary:
    """Heuristic summary of a session for listings and history UIs.

    Attributes:
        session_id: Source session id.
        title: Short title derived from the first prompt.
        synopsis: One-paragraph summary of what happened.
        key_actions: Distinct tool names invoked, in first-seen order.
        outcome: ``"success"``, ``"failure"``, ``"mixed"`` or ``"unknown"``.
        turn_count: Number of turns in the session.
        agent_ids: Distinct agent ids that produced turns.
        char_count: Total characters across all prompts and responses.
    """

    session_id: str
    title: str = ""
    synopsis: str = ""
    key_actions: list[str] = field(default_factory=list)
    outcome: str = "unknown"
    turn_count: int = 0
    agent_ids: list[str] = field(default_factory=list)
    char_count: int = 0

    def to_dict(self) -> dict[str, tp.Any]:
        """Return a JSON-ready dict of the summary fields."""

        return asdict(self)


class SessionSummarizer:
    """Generates a :class:`SessionSummary` from a :class:`SessionRecord`.

    Pure-Python heuristics by default; an optional ``llm_client`` callable
    may refine the synopsis after the draft is computed.
    """

    def __init__(
        self,
        llm_client: tp.Callable[[str], str] | None = None,
    ) -> None:
        """Configure the summarizer with an optional LLM refinement hook.

        Args:
            llm_client: Callable that takes a prompt and returns the refined
                synopsis. Errors are caught and the heuristic draft is kept.
        """

        self.llm_client = llm_client

    def summarize(self, session: SessionRecord) -> SessionSummary:
        """Produce a :class:`SessionSummary` for ``session``.

        Heuristics run first; if an ``llm_client`` was supplied it gets a
        chance to rewrite the synopsis. LLM exceptions are swallowed so a
        flaky model never breaks listings.
        """

        title = self._derive_title(session)
        synopsis = self._derive_synopsis(session)
        key_actions = self._collect_tools(session)
        outcome = self._derive_outcome(session)
        agents = self._distinct_agents(session)
        char_count = sum(len(t.prompt or "") + len(t.response_content or "") for t in session.turns)
        if self.llm_client is not None and session.turns:
            try:
                synopsis = self._refine_with_llm(session, synopsis) or synopsis
            except Exception:
                logger.debug("LLM refinement failed", exc_info=True)
        return SessionSummary(
            session_id=session.session_id,
            title=title,
            synopsis=synopsis,
            key_actions=key_actions,
            outcome=outcome,
            turn_count=len(session.turns),
            agent_ids=agents,
            char_count=char_count,
        )

    def _derive_title(self, session: SessionRecord) -> str:
        """Derive a short title from the first user prompt."""

        if not session.turns:
            return f"Session {session.session_id[:8]}"
        first = session.turns[0]
        prompt = (first.prompt or "").strip()
        if not prompt:
            return f"Session {session.session_id[:8]}"
        words = prompt.split()
        if len(words) > 12:
            return " ".join(words[:10]) + "…"
        return prompt[:80]

    def _derive_synopsis(self, session: SessionRecord) -> str:
        """Compose a 1-3 sentence synopsis from prompt, tool count, response."""

        if not session.turns:
            return "Empty session."
        first_prompt = (session.turns[0].prompt or "").strip()
        last_response = ""
        for t in reversed(session.turns):
            if t.response_content:
                last_response = t.response_content.strip()
                break
        n_tools = sum(len(t.tool_calls) for t in session.turns)
        sentences: list[str] = []
        if first_prompt:
            sentences.append(f'User asked: "{_truncate(first_prompt, 120)}".')
        if n_tools:
            sentences.append(f"Agent used {n_tools} tool call(s) across {len(session.turns)} turn(s).")
        else:
            sentences.append(f"Agent answered in {len(session.turns)} turn(s) without tools.")
        if last_response:
            sentences.append(f'Final answer: "{_truncate(last_response, 200)}".')
        return " ".join(sentences)

    def _collect_tools(self, session: SessionRecord) -> list[str]:
        """Return distinct tool names invoked, preserving first-seen order."""

        seen: set[str] = set()
        out: list[str] = []
        for turn in session.turns:
            for call in turn.tool_calls:
                name = getattr(call, "tool_name", None) or getattr(call, "name", "")
                if name and name not in seen:
                    seen.add(name)
                    out.append(name)
        return out

    def _derive_outcome(self, session: SessionRecord) -> str:
        """Classify overall outcome from the per-turn ``status`` values."""

        if not session.turns:
            return "unknown"
        statuses = [t.status for t in session.turns]
        if all(s == "success" for s in statuses):
            return "success"
        if all(s != "success" for s in statuses):
            return "failure"
        return "mixed"

    def _distinct_agents(self, session: SessionRecord) -> list[str]:
        """Return distinct agent ids that produced turns, in first-seen order."""

        seen: set[str] = set()
        out: list[str] = []
        for turn in session.turns:
            if turn.agent_id and turn.agent_id not in seen:
                seen.add(turn.agent_id)
                out.append(turn.agent_id)
        return out

    def _refine_with_llm(self, session: SessionRecord, draft: str) -> str:
        """Ask the configured LLM to rewrite ``draft`` more concisely."""

        prompt = (
            "Rewrite this session synopsis as 1-3 short, neutral sentences. "
            "Preserve all factual claims; do not invent details.\n\n"
            f"Draft:\n{draft}\n\n"
            "Recent turns (newest last):\n"
            + "\n".join(
                f"- USER: {(t.prompt or '')[:120]} | AGENT: {(t.response_content or '')[:120]}"
                for t in session.turns[-3:]
            )
        )
        if self.llm_client is None:
            return ""
        out = self.llm_client(prompt)
        return out.strip() if isinstance(out, str) else ""


def _truncate(text: str, n: int) -> str:
    """Collapse whitespace and trim ``text`` to at most ``n`` chars."""

    text = " ".join(text.split())
    if len(text) <= n:
        return text
    return text[: n - 1].rstrip() + "…"


__all__ = ["SessionSummarizer", "SessionSummary"]

File: src/typescript/src/session/summarizer.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { SessionRecord } from './models.js'

export interface SessionSummaryOptions {
  readonly agentIds?: readonly string[]
  readonly charCount?: number
  readonly keyActions?: readonly string[]
  readonly outcome?: 'failure' | 'mixed' | 'success' | 'unknown'
  readonly sessionId: string
  readonly synopsis?: string
  readonly title?: string
  readonly turnCount?: number
}

/** Cheap, factual session summary suitable for history listings. */
export class SessionSummary {
  agentIds: string[]
  charCount: number
  keyActions: string[]
  outcome: 'failure' | 'mixed' | 'success' | 'unknown'
  sessionId: string
  synopsis: string
  title: string
  turnCount: number

  constructor(options: SessionSummaryOptions) {
    this.sessionId = options.sessionId
    this.title = options.title ?? ''
    this.synopsis = options.synopsis ?? ''
    this.keyActions = [...options.keyActions ?? []]
    this.outcome = options.outcome ?? 'unknown'
    this.turnCount = options.turnCount ?? 0
    this.agentIds = [...options.agentIds ?? []]
    this.charCount = options.charCount ?? 0
  }

  toRecord(): Record<string, unknown> {
    return {
      session_id: this.sessionId,
      title: this.title,
      synopsis: this.synopsis,
      key_actions: [...this.keyActions],
      outcome: this.outcome,
      turn_count: this.turnCount,
      agent_ids: [...this.agentIds],
      char_count: this.charCount,
    }
  }
}

export interface SessionSummarizerOptions {
  readonly llmClient?: (prompt: string) => string
}

/** Produces a deterministic heuristic summary, optionally refined by a caller-owned LLM. */
export class SessionSummarizer {
  private readonly llmClient: ((prompt: string) => string) | undefined

  constructor(options: SessionSummarizerOptions = {}) {
    this.llmClient = options.llmClient
  }

  summarize(session: SessionRecord): SessionSummary {
    const draft = this.deriveSynopsis(session)
    let synopsis = draft
    if (this.llmClient && session.turns.length > 0) {
      try {
        synopsis = this.refine(session, draft) || draft
      } catch {
        synopsis = draft
      }
    }
    return new SessionSummary({
      sessionId: session.sessionId,
      title: this.deriveTitle(session),
      synopsis,
      keyActions: this.collectTools(session),
      outcome: this.deriveOutcome(session),
      turnCount: session.turns.length,
      agentIds: this.distinctAgents(session),
      charCount: session.turns.reduce(
        (count, turn) => count + turn.prompt.length + (turn.responseContent?.length ?? 0),
        0,
      ),
    })
  }

  private collectTools(session: SessionRecord): string[] {
    const seen = new Set<string>()
    const tools: string[] = []
    for (const turn of session.turns) {
      for (const call of turn.toolCalls) {
        if (call.toolName && !seen.has(call.toolName)) {
          seen.add(call.toolName)
          tools.push(call.toolName)
        }
      }
    }
    return tools
  }

  private deriveOutcome(session: SessionRecord): 'failure' | 'mixed' | 'success' | 'unknown' {
    if (session.turns.length === 0) return 'unknown'
    if (session.turns.every(turn => turn.status === 'success')) return 'success'
    if (session.turns.every(turn => turn.status !== 'success')) return 'failure'
    return 'mixed'
  }

  private deriveSynopsis(session: SessionRecord): string {
    if (session.turns.length === 0) return 'Empty session.'
    const firstPrompt = session.turns[0]?.prompt.trim() ?? ''
    const lastResponse = [...session.turns].reverse().find(turn => turn.responseContent)?.responseContent?.trim() ?? ''
    const toolCount = session.turns.reduce((count, turn) => count + turn.toolCalls.length, 0)
    const sentences: string[] = []
    if (firstPrompt) sentences.push(`User asked: "${truncate(firstPrompt, 120)}".`)
    sentences.push(
      toolCount > 0
        ? `Agent used ${toolCount} tool call(s) across ${session.turns.length} turn(s).`
        : `Agent answered in ${session.turns.length} turn(s) without tools.`,
    )
    if (lastResponse) sentences.push(`Final answer: "${truncate(lastResponse, 200)}".`)
    return sentences.join(' ')
  }

  private deriveTitle(session: SessionRecord): string {
    const prompt = session.turns[0]?.prompt.trim()
    if (!prompt) return `Session ${session.sessionId.slice(0, 8)}`
    const words = prompt.split(/\s+/)
    if (words.length > 12) return `${words.slice(0, 10).join(' ')}…`
    return prompt.slice(0, 80)
  }

  private distinctAgents(session: SessionRecord): string[] {
    const seen = new Set<string>()
    const agents: string[] = []
    for (const turn of session.turns) {
      if (turn.agentId && !seen.has(turn.agentId)) {
        seen.add(turn.agentId)
        agents.push(turn.agentId)
      }
    }
    return agents
  }

  private refine(session: SessionRecord, draft: string): string {
    if (!this.llmClient) return ''
    const recent = session.turns.slice(-3)
      .map(turn => `- USER: ${turn.prompt.slice(0, 120)} | AGENT: ${(turn.responseContent ?? '').slice(0, 120)}`)
      .join('\n')
    return this.llmClient(
      'Rewrite this session synopsis as 1-3 short, neutral sentences. Preserve all factual claims; do not invent details.\n\n'
        + `Draft:\n${draft}\n\nRecent turns (newest last):\n${recent}`,
    ).trim()
  }
}

function truncate(value: string, length: number): string {
  const collapsed = value.split(/\s+/).filter(Boolean).join(' ')
  return collapsed.length <= length ? collapsed : `${collapsed.slice(0, length - 1).trimEnd()}…`
}
